env/io_posix.cc

```diff
@@ -1686,15 +1702,19 @@ IOStatus PosixDirectory::FsyncWithDirOptions(
     }
     // fallback to dir-fsync for kDefault, kDirRenamed and kFileDeleted
   }
+
+  // skip fsync/fcntl when fd_ == -1 since this file descriptor has been closed
+  // in either the de-construction or the close function, data must have been
+  // fsync-ed before de-construction and close is called
```

Comment:
> data must have been // fsync-ed before de-construction and close is called
I'm actually not quite familiar with this - is there any posix rule about this or a contract we impose on callers of `PosixDirectory::close()`?

Based on a quick search, I get 
```
A successful close does not guarantee that the data has been
       successfully saved to disk, as the kernel uses the buffer cache
       to defer writes.  Typically, filesystems do not flush buffers
       when a file is closed.  If you need to be sure that the data is
       physically stored on the underlying disk, use [fsync(2)](https://man7.org/linux/man-pages/man2/fsync.2.html).
``` 
from https://man7.org/linux/man-pages/man2/close.2.html,  which reads quite the opposite (?).

Another question I have is `fd_ = -1;` happens even when `close()` fail. So even if such rule or contract exists (as an answer to my question above), can `fd_==1` still safely imply `fsync/fcntl` is called?

Reply:
Thanks for pointing out that. Yes, you are right. Should I move fd_ = -1 into the condition branch when close succeeds?

Reply:
I am thinking if we should fsync or fcntl inside close() and remove fd_ = -1 as you did - any thoughts on this @ajkr ?

Reply:
Thanks for your reviewing. I'm pushing another version with suggested comment added.